site/source/_posts/2017/11/17-Dojo-2-Beta-4/Dojo-2-Beta-4.md

```diff
@@ -0,0 +1,263 @@
+    }
+
+    protected onDetach() {
+        // perform actions when detached
+    }
+
+    protected render() {
+        return v('div', { }, [ 'Hello world!' ])
+    }
+}
+```
+
+Previously, we had other capabilities which provided functionality in this area, but we have deprecated those and will remove them in the future.
+
+## Deferred properties
+
+A virtual DOM node can now provide its properties in a way that are resolved in a deferred fashion, where instead of providing an object of `VirtualDomProperties`, you can provide a function which takes a boolean flag of if the virtual DOM node has been inserted into the DOM and returns an object of `VirtualDomProperties`.
+
+Because these functions will be called during the next `requestAnimationFrame` (when the projector is running in its default asynchronous mode), this feature can be used to better handle CSS transition states.  For example, you can now do something like this:
+
+```typescript
+import { v } from '@dojo/widget-core/d';
+import WidgetBase from '@dojo/widget-core/WidgetBase';
+
+render() {
+    return v('div', (inserted) => {
+        return {
+            classes: 'fadeInTransition',
+            styles: {
+                opacity: inserted ? '1' : '0'
+            }
+        };
+    }, [ 'Hello World!' ]);
+}
+```
+
+## Web animations meta provider
+
+One of the goals of the Dojo 2 widgeting system is to provide sufficient tools and abstractions so that widget authors don't have to deal directly with the DOM.  Meta providers provide ways to interact with the DOM, sometimes in complex ways, without having to sweat the details.
+
+In this vein, we have added a web animations meta provider in Beta 4.  It provides a way to easily animate DOM nodes in a reactive way, as just part of your render.  For example:
+
+```typescript
+import { v } from '@dojo/widget-core/d';
+import WidgetBase from '@dojo/widget-core/WidgetBase';
+import WebAnimation from '@dojo/widget-core/meta/WebAnimation';
+
+export default class AnimatedWidget extends WidgetBase {
+    protected render() {
+        const animate = {
+            id: 'rootAnimation',
+            effects: [
+                { height: '10px' },
+                { height: '100px' }
+            ],
+            controls: {
+                play: true
+            }
+        };
+
+        this.meta(WebAnimation).animate('root', animate);
+
+        return v('div', {
+            key: 'root'
+        });
+    }
+}
+```
+
+## Custom Elements/Web Components
+
+One of the main objectives for Dojo 2 was that there was a high level of interoperability with Web Components, both using Web Components in Dojo 2 applications but also being able to use Dojo 2 widgets easily as Custom Elements.  We made several changes to the widgeting system to better support Dojo 2 widgets used as Custom Elements, specifically making it efficient to use children Custom Elements of a Dojo 2 widget Custom Element.
+
+So what you would write in TypeScript:
```

Comment:
what you would write in TypeScript: -> If you write the following in TypeScript: